ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala

```diff
@@ -996,39 +996,52 @@ private final class OracleQueries(
   private[this] def pathSteps(path: JsonPath): Cord =
     path.elems.foldMap(_.fold(k => (".\"": Cord) ++ k :- '"', (_: _0.type) => "[0]"))
 
-  // I cannot believe this function exists in 2021
-  // None if literal is too long
-  // ORA-40454: path expression not a literal
-  private[this] def oraclePathEscape(readyPath: Cord): Option[Fragment] = for {
-    s <- Some(readyPath.toString)
-    if s.size <= literalStringSizeLimit
-    _ = assert(
+  // Takes a literal text value and expresses it as a single-quoted value which can be embedded
+  // within json queries.
+  private[this] def oracleJsonStringEscape(readyPath: Cord): Fragment = {
+    val s = readyPath.toString
+    assert(
       !s.startsWith("'") && !s.endsWith("'"),
       "Oracle JSON query syntax doesn't allow ' at beginning or ending",
     )
-    escaped = s.replace("'", "''")
-    if escaped.length <= literalStringSizeLimit
-  } yield Fragment const0 ("'" + escaped + "'")
+    val escaped = s.replace("'", "''")
+    Fragment.const0("'" + escaped + "'")
+  }
```

Comment:
This function was written with two things in mind:

1. the oracle JSON path-query syntax (hence the original name), and
2. allowed Daml-LF record/variant field names

For example, the `assert` is not acceptable for the new use case you're assigning to this function, because while generated Oracle json queries will never start or end with `'`, there is no reason a user-supplied function wouldn't. The `assert` _was_ just a sanity check, but _now_ it's taking on the job of preventing an injection attack. In this specific case you should still pass as a query parameter rather than failing an assertion.

Additionally, I believe that the literal string size limit still applies. Passing as a query parameter is how to deal with this problem when contract payload indexing is off, but I think if you use this fallback here you'll just hit #10780 (note the `query record contains handles` tests are also guarded with `onlyIfLargeQueries`).

My overall suggestions:

* [ ] consider whether any characters other than `'` need to be escaped; `'` was the only important one with JSON path query syntax but now all characters are possible
* [ ] review the function carefully for potential SQL injection vectors
* [ ] add a scalacheck property check to the db-backend tests that applies the escape function to random strings and ensures that the string literal doesn't terminate until the last character, allowing injection (this can be annotated as a security test). You've definitely covered an important case, but I'm also worried about injection attacks that have not been documented in xkcd 🙂 
* [ ] take the assertion path for `'`-checking when escaping a JSON path query, but fall back to a query parameter or another approach if a user-supplied scalar begins or ends with `'`
* [ ] large strings are sort of up in the air; either do a query parameter fallback if you break `literalStringSizeLimit`, or throw an error in a similar vein as [our existing documentation of this issue](https://docs.daml.com/2.6.1/json-api/search-query-language.html#when-using-oracle-queries-fail-if-a-token-is-too-large). I believe the limit applies to the post-escaped form, for what it's worth

Reply:
Thanks Stephen, these are some good points.

I've reverted `oracleShortPathEscape` and `oraclePathEscape` to be untouched, and added a new function `quotedJsonSearchToken` to serve this new purpose. It is also aware of the size limit - 256, which I've added tests for.

As far as I can tell, the the same escaping should work in principle. I did add unit and property tests for this escaping function, but I think the truest test of how it is interpreted would be in running them against the real database. I've done that for a couple more interesting cases.